ug: add checkpoints tutorial

flippedcoder opened the pull request:
This adds some more detailed documentation around the checkpoints implementation in training models.

dmpetrov commented on content/docs/start/checkpoints.md:
```diff
@@ -0,0 +1,163 @@
+`ini.txt` file.
+
+If you want to start from a specific existing checkpoint, you'll need to run
+`dvc exp run --rev 123455` where `12345` is the checkpoint ID for specific
```
It seems like `--rev` works only for queue/temp runs, not for this scenario. Created an issue: https://github.com/iterative/dvc/issues/5814

dmpetrov commented on content/docs/start/checkpoints.md:
```diff
@@ -0,0 +1,163 @@
+These experiments have IDs associated with each of the checkpoints so you have a
+reference to each epoch.
+
+## Starting from an existing checkpoint
```
I'd start with continue\resuming training: train for a few epochs, terminate and continue training back. It supposes to create a straight line of metrics improvements.

dmpetrov commented on content/docs/start/checkpoints.md:
```diff
@@ -0,0 +1,163 @@
+│ │ ╟ 9b0bf16   │ 01:20 PM │    2 │ 0.9271 │
+│ │ ╟ 39d1444   │ 01:19 PM │    1 │ 0.9157 │
+│ ├─╨ fcefafb   │ 01:19 PM │    0 │ 0.8572 │
+└───────────────┴──────────┴──────┴────────┘
```
It seems like the algorithm is too good for showing the value of checkpoints 😄  You are getting 85% accuracy in 1st epoch and 95% in 4th. You might consider using a weaker network architecture or a less aggressive learning rate to show a smoother dynamic.

dmpetrov commented on content/docs/start/checkpoints.md:
```diff
@@ -0,0 +1,163 @@
+this: `dvc exp run --reset`.
+
+This removes the `model.pt` file and clears the `dvc.lock` file of any existing
+checkpoints.
```
I'd add apply scenario: `dvc exp apply && modify file && dvc exp run`. It might be very useful.

flippedcoder replied:
What does `modify file` do here? Does that remove the `model.pt` file since there's no checkpoint id in the apply command?

dmpetrov replied:
Modify means - make changes in params.yaml or train.py

dmpetrov reviewed (changes requested):
@flippedcoder an amazing tutorial ✨ 
I left a few comments inline. 

PS: you can take some ideas from here https://www.youtube.com/watch?v=J8mCr3wVgdA

jorgeorpinel commented:
Thanks for this contribution @flippedcoder! Are we sure we want to expand the Get Started with this somewhat specific guide (cc @shcheklein)? If not, we can definitely find another place to put this info, and link from https://dvc.org/doc/start/experiments.

Happy to help here once that is decided and @dmitry's comments are addressed.

dberenbaum commented:
Thanks @flippedcoder! This looks awesome, and it helps address an open issue: https://github.com/iterative/dvc.org/issues/2292.

One overarching comment is that we might need to think about where this fits. Having a get started page for checkpoints is important, but appending it to the existing get started section might be a bit confusing at the moment since all the other pages are iterations of the same scenario in https://github.com/iterative/example-get-started and this is a completely new scenario. I think there is some ongoing work to maybe introduce an MNIST version of all the get started pages, but I'm not sure the status. What do you think @dmpetrov @shcheklein?

Also, the existing https://github.com/dberenbaum/dvc-checkpoints-mnist repo has different branches for a bunch of different scenarios:
* basic: This is the most basic way to use checkpoints, creating a single checkpoint after the script completes. This is not an expected typical workflow, but it's arguably the simplest way to introduce the concept without having to modify the underlying script.
* make_checkpoint: This introduces how to generate multiple checkpoints within a script using `make_checkpoint()`.
* signal_file: This introduces a language-agnostic way to mimic `make_checkpoint` for non-Python projects.
* live: This integrates `make_checkpoint()` with https://github.com/iterative/dvclive.
* full_pipeline: This adds more general dvc complexity, like multiple stages and plots. It could probably be ignored or deleted unless we develop a use case that discusses integrating those dvc features with this checkpoints example.

We probably need just one of these branches for get started. I have the `live` branch as the default because it is probably the most natural and fully featured workflow, but `make_checkpoint` is similar and doesn't have the complication of the `dvclive` dependency. We could even start with `basic` and then add `make_checkpoint` functionality (which looks like kind of what you did here).

dmpetrov commented:
@jorgeorpinel & @dberenbaum it is not mentioned here but it is a Draft and the place/url for this doc is not defined yet. What would be the best place to put this document? A new Deep Learning Checkpoints section in Use Cases?

dberenbaum commented:
> @jorgeorpinel & @dberenbaum it is not mentioned here but it is a Draft and the place/url for this doc is not defined yet. What would be the best place to put this document? A new Deep Learning Checkpoints section in Use Cases?

Right now, that seems like a good spot for it, but we are also working on developing a docs roadmap, so we might want to reconsider once we have that.

dmpetrov commented:
> working on developing a docs roadmap, so we might want to reconsider once we have that.

Right. But even if we found a better use case for checkpoints it still makes sense to use the Mnist use case as a temporary solution.

jorgeorpinel commented:
> helps address an open issue: #2292

The current content doesn't seem to address that, but maybe there should be a section for it?

> the existing https://github.com/iterative/dvc-checkpoints-mnist repo has different branches for a bunch of different scenarios

Agree that it would be great to use that here (the one under `iterative/`). I'm working on adapting other existing examples to use it too; Also that this content should focus on the main workflow (`make_checkpoint`) + mention other possibilities, to keep it short.

> What would be the best place to put this document?

The content looks like a guide to me. One option is to help @flippedcoder compact this a bit and put it under https://dvc.org/doc/user-guide/experiment-management (which prob should be a section anyway — separate issue rel #2269 and #2367).

jorgeorpinel commented:
> working on developing a docs roadmap

BTW this is the first PR in https://github.com/iterative/dvc.org/projects/1 😬

dberenbaum commented:
> the one under `iterative/`

Whoops, that was a mistake. I edited the comment above with the correct link. By the way, I think @flippedcoder is already using that and probably just needs to link to it in the doc.

shcheklein commented:
> This content looks like a guide to me. One option is to help @flippedcoder compact this a bit and put it under https://dvc.org/doc/user-guide/experiment-management

I like this idea! (+ other suggestion to make precise, reuse MNIST if possible, probably make it actionable). This place is indeed a better fit for now. In the get started Experiments section we should put a link for now to the UG section.

@flippedcoder congrats with the first PR! Awesome work 🎉

dmpetrov commented:
> it helps address an open issue: #2292.

@dberenbaum I was thinking it is a bit separate use case, not the basic one. Do you think it makes sense to combine these two?

@flippedcoder you can keep working on the doc here and then move it under Use Cases. Or move it now. Ups to you 😄

jorgeorpinel commented:
> move it under Use Cases

User Guide right? 🤔 BTW to put it in /user-guide/experiment-management/checkpoints (which I'm suggesting) we'd need other small changes. For now it can be placed directly in /doc/user-guide/checkpoints (below **Experiment Management**) I think. And I can jump in to finalize the structure before merging.

dberenbaum commented:
> > it helps address an open issue: #2292.
> 
> @dberenbaum I was thinking it is a bit separate scenario, not the basic one. Do you think it makes sense to combine these two?

It's fine if it doesn't address #2292. In my mind, that had become a placeholder for adding more robust checkpoints docs.

---

What's the purpose of this doc? We have needs for multiple checkpoint docs IMO:
* For getting started, agree with @shcheklein that a link is probably enough (for now).
* A use case would be great to show how checkpoints could be used in a typical deep learning project workflow and what value they add. This PR could do that, or the use case could be an opportunity for @flippedcoder to come up with a more interesting project to flesh out more.
* A user guide would also be great to explain how checkpoints work. This PR could also do that. It seems like one suggestion is to compact it and focus on a single workflow like `make_checkpoint`. I think we might want to start there but also explain the different scenarios in the branches of the repo in the user guide. If there are ones that aren't helpful, we can exclude them, but the `make_signal` scenario for example seems necessary to me.

By the way, I think this confusion is my own fault for giving this idea to @dmpetrov and not providing enough context to him or @flippedcoder, so sorry about that!

jorgeorpinel commented:
> use case to show how checkpoints could be used in a typical deep learning project workflow

I'm not against this idea. It's just that the drafted content so far looks more like an explanation of the checkpoints feature (a DVC guide). Maybe partially the result of @flippedcoder's learning process? (Great practice to write in those situations BTW.)

dberenbaum commented:
> > use case to show how checkpoints could be used in a typical deep learning project workflow
> 
> I'm not against this idea. It's just that the drafted content so far looks more like an explanation of the checkpoints feature (a DVC guide). Maybe partially the result of @flippedcoder's learning process? (Great practice to write in those situations BTW.)

Yup, seems like there's consensus to make this a user guide. What about addressing the scenarios in the different repo branches?

flippedcoder commented:
Hey everyone! Thanks for all the great feedback! I moved it from the Getting Started section to the User Guide section with the updates @dmpetrov mentioned. I'm working on making the algorithm perform worse to show off checkpoints better. 😅 Otherwise, I'm open to any other feedback or suggestions y'all have.

jorgeorpinel commented:
> What about addressing the scenarios in the different repo branches?

For now I'd just consider mentioning them somewhere in this PR as a bullet list or something or not even that and make a separate issue perhaps?

Cool @flippedcoder thanks! So is this ready for review? 🙂 (no longer draft status)

![image](https://user-images.githubusercontent.com/1477535/114950532-e6972180-9e18-11eb-8f0d-070269103422.png)

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+When you need to completely retrain a model and you don't want to use any of the
+existing checkpoints, you can get rid of all of them with the `--reset` flag.
+Add this to the usual command for running DVC experiments and it looks like
+this: `dvc exp run --reset`.
```
It would be great to see the table when you can point to the metrics that start from zero, not form the previous value.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+this: `dvc exp run --reset`.
+
+You can also reset checkpoints with the following command:
+`dvc exp apply && modify file && dvc exp run`.
```
That's not clear 😄  What needs to be modified? Removing `checkpoint:true` from `dvc.yaml`? If so, it is a special case and probably should be in the basic tutorial.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+to save the model files at checkpoints during a training process and return back
+when metrics start diverging.
+
+_Note: This is not a continuation of the previous Getting Started sections._
```
We moved it from Getting started. It seems like this is not needed anymore.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+ML checkpoints are an important part of deep learning because ML engineers like
+to save the model files at checkpoints during a training process and return back
+when metrics start diverging.
+
```
Some ideas that can be used in the intro: "... saving ML model weights as checkpoints is one of the best practices in deep learning projects... DVC automate this best practice ... and help users not only to have the history of checkpoints but also keep track of all code and data changes corresponded to the checkpoints ... DVC introduces reproducibility to a messy deep learning process"

dmpetrov replied:
Also, we should mention how checkpoints are implemented in DVC (a separate 1-2 paragraph section). We use Git commits, but these are not regular commits and branches, these are "ephemeral" commits and branches that not over pollute your branches' workspace. After you you are happy with some of the experiments or checkpoints you can "materialize" an "ephemeral" experiment branch.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+`dvc exp apply && modify file && dvc exp run`.
+
+This removes the `model.pt` file and clears the `dvc.lock` file of any existing
+checkpoints.
```
Subsections that might be added (for E2E tutorial):
1. Navigate through metrics and plots: `dvc metrics diff exp-12345 abc543` and plots `dvc plots diff deca134 fea9483`. We should emphasize compatibility with other DVC features and "commutative effect".
2. Nevigate in code - `git diff deca134 fea9483 -- train.py params.yaml`. We should emphasize compatibility with Git (not DVC) and "commutative effect". A reference to DevOps and MLOps would be great.
3. How to "materialize" an "ephemeral" experiment branch - `dvc exp branch`.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,195 @@
+│ │ ╟ 39d1444   │ 01:19 PM │    1 │ 0.9157 │
+│ ├─╨ fcefafb   │ 01:19 PM │    0 │ 0.8572 │
+└───────────────┴──────────┴──────┴────────┘
+```
````
It might be better to show an actual table that the user will get after the 1st run - the metrics will start from something like 0.10 in the 1st epoch, not 0.85.

jorgeorpinel replied:
Also this table looks to be repeated in the **Viewing checkpoints** section. Probably just once is OK since there's another `exp show` example after `exp apply`?

dberenbaum replied:
@dmpetrov The first iteration used to be a lot worse because it was using the full batch instead of minibatch. I primarily switched to minibatch to speed up the checkpoints, but it also made performance much better. Maybe in between would be good, since the old example took a bit too long to get decent results and tended to sometimes even get worse between checkpoints. The architecture is already pretty simple:

```
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
================================================================
            Conv2d-1            [-1, 8, 28, 28]              80
         MaxPool2d-2            [-1, 8, 14, 14]               0
            Conv2d-3           [-1, 16, 14, 14]           1,168
            Linear-4                   [-1, 32]         100,384
            Linear-5                   [-1, 10]             330
================================================================
Total params: 101,962
Trainable params: 101,962
Non-trainable params: 0
----------------------------------------------------------------
Input size (MB): 0.00
Forward/backward pass size (MB): 0.08
Params size (MB): 0.39
Estimated Total Size (MB): 0.48
----------------------------------------------------------------
```

We can probably make it even simpler (take out the second convolutional layer, reduce the number of channels even further) if you think it's "too easy" as is. Or we could switch to a different dataset (the recent dolt-dvc blog post had a good idea to use fashion-mnist as a harder alternative).

dberenbaum replied:
@flippedcoder Maybe try out the different MNIST variants in https://pytorch.org/vision/stable/datasets.html? I think they all should be close to drop-in replacements but not as easy as the original MNIST.

dberenbaum replied:
Sorry, I wasn't clear in my first response.

> the metrics will start from something like 0.10 in the 1st epoch, not 0.85.

What I meant is that this is not true now. The current repo jumps immediately to 0.85 after one epoch.

dmpetrov reviewed (changes requested):
Great change @flippedcoder! Also, I'm glad that we found the place for this doc in user-guide 😄 

I put mane comments mostly because of my comments based on the assumption that we are making an end-to-end tutorial, not a simple getting started that was an initial task.

@jorgeorpinel @dberenbaum does it make sense to extend it to E2E or we should keep it as a simple getting-stared?

flippedcoder commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+
+You can also add checkpoints by running the command
+`dvc stage add -n name_of_stage --checkpoints model.pt`. This will create a new
+stage with checkpoints enabled.
```
So if they have an existing stage, the only way to add checkpoints is to manually update the `dvc.yaml` with `checkpoint: true`?

jorgeorpinel replied:
You can rewrite stages with `stage add -f`. But anyway, what about making this a very shot note about `stage add` since I think we want to encourage users to edit dvc.yaml manually? Something like

"See also `dvc stage add --checkpoint` to generate the stage from CLI."

jorgeorpinel commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+
+When you want to add a checkpoint to your training process, there are only a
+couple of things you need to setup. You can find the code we use here in
+[this GitHub repo](https://github.com/iterative/dvc-checkpoints-mnist).
```
> we found the place for this doc in user-guide
> does it make sense to extend it to E2E or we should keep it as a simple getting-stared?

To clarify, the User Guide is (meant to be) even simpler that even Get Started (see [longer ref](https://documentation.divio.com/explanation/#explanation)). Meaning even further from E2E (no continuity needed).

I'd say keep the current plan and merge it as a guide, and THEN consider turning it into a walkthrough/tutorial/we'll-see. Just to keep this review manageable.

jorgeorpinel commented on content/docs/sidebar.json:
```diff
@@ -134,6 +134,12 @@
           "merge-conflicts"
         ]
       },
+      {
+        "slug": "checkpoints",
+        "tutorials": {
+          "katacoda": "https://katacoda.com/dvc/courses/get-started/checkpoints"
+        }
+      },
       "experiment-management",
```
Let's put this right after the main Exps guide? Committable suggestion:
```suggestion
      "experiment-management",
      {
        "slug": "checkpoints",
        "tutorials": {
          "katacoda": "https://katacoda.com/dvc/courses/get-started/checkpoints"
        }
      },
```

jorgeorpinel commented on content/docs/start/index.md:
```diff
@@ -76,3 +76,8 @@ by one in the next few pages:
   comparison across many ML experiments. Track your experiments with automatic
   versioning and checkpoint logging. Compare differences in parameters, metrics,
   code, and data. Apply, drop, roll back, resume, or share any experiment.
+
+- [**Checkpoints**](/doc/user-guide/checkpoints) let you track every training
+  epoch, return to any previous checkpoint, see when metrics converge or when
+  the learning rate needs to be adjusted, and even recover model weights,
+  parameters, and code.
```
No longer needed in the Get Started index (if it's going to be a guide). But we can link from (and add a little text if needed) in https://dvc.org/doc/user-guide/experiment-management#checkpoints-in-source-code

jorgeorpinel commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,195 @@
+have the ability to resume training from previous checkpoints. When you adjust
+parameters and code, DVC tracks those changes for you.
+
+## Adding checkpoints
```
What about

1. Enabling DVC pipelines for checkpoint experiments
2. Registering checkpoints in your code

> That terminology would be consistent with existing docs.

jorgeorpinel commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,195 @@
+```
+
+_We don't use `dvc repro` for checkpoints because it doesn't know how to handle
+checkpoints outputs and it will run the stage from scratch._
````
I think `dvc repro` will give an error with checkpoint-enabled pipelines, please double check using the latest DVC version, if you haven't.

jorgeorpinel commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,195 @@
+```python
+if i_ % 5 == 0:
+    make_checkpoint()
+```
````
You can use the `git diff` format (` ```git ` in md). See for example in https://dvc.org/doc/command-reference/exp/apply#example-make-an-experiment-persistent:

![image](https://user-images.githubusercontent.com/1477535/115304592-b6a69180-a12a-11eb-9741-3ca8aa5c4430.png)

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,386 @@
+you can open in a browser and it will display plots for you similar to the ones
+below.
+
+![Plots generated from running experiments on MNIST dataset using DVC](/img/checkpoints_plots.png)
```
The same as metrics diff - it would be great to see the actual command. Also, with the simple `acc` metrics the plot diff might not look impressive. We can think about a better plot for metrics.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,386 @@
+
+## Adding checkpoints to your Git workspace
+
+When the final epoch has run, you'll see a few commands in the terminal that
```
`final epoch has` - terminating a job seems like a more common scenario. "After terminating the training ..."?

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,386 @@
+
+![Plots generated from running experiments on MNIST dataset using DVC](/img/checkpoints_plots.png)
+
+## Adding checkpoints to your Git workspace
```
"Git workspace" might be not the best term here. "Adding checkpoints to Git" might be enough.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,386 @@
+dvc exp apply exp-263da
+git add .
+git commit -m 'optimize model'
+```
````
Do we need `apply` here? I see two possible actions (from the output):
1. `git add dvclive.json dvc.yaml .gitignore train.py dvc.lock && git commit -m 'my experiment'` to make a single commit in the current branch (like master).
2. `dvc exp branch <exp>` to create a branch that can be stored for later or pushed to GitHub/CiCd.

The first seems like an easy option and we can provide a code sample for a user to run.

dmpetrov reviewed (changes requested):
That's a great improvement. Now it is a tutorial!

Some changes are inline. It seems like the default repo does not work well and a new one should be created for a smoother user experience. What needs to be changed:
1. Ideally, the code from the cloned repo should work through `python train.py` but with no checkpoints.
2. User should generate the stage `dvc stage add -n name --checkpoint ... python ...`
3. User should add metrics reporting and checkpoints to code
4. We might need some summary plots to make `plots diff` meaningful. 

Happy to answer any questions.

dberenbaum commented:
I haven't had a chance to read through this doc yet, but I have a few comments from skimming through it:

* There is a similar MNIST project repo under development to update the interactive getting started examples. See https://github.com/iterative/katacoda-scenarios/issues/60. The underlying MNIST projects are pretty similar, so it might make sense to work towards a single MNIST repo.
* Collaboration between these projects might also help clarify what to include in getting started (likely one very simple checkpoints scenario) vs. this type of in-depth user guide (which might include all the ways to add checkpoints, how to navigate around checkpoint experiments, and maybe how they can be combined with parameter tuning).
* As others have mentioned, we will need a more polished repo with different tags/entrypoints along the way, and the docs should detail how to get from one point to the next. The current getting started in https://dvc.org/doc/start does this using tags in https://github.com/iterative/example-get-started, and that repo and its tags are generated from https://github.com/iterative/example-repos-dev.
* I need to clarify the order of branches in https://github.com/iterative/dvc-checkpoints-mnist. For tutorial purposes, it would make sense to go from no checkpoints -> `basic` (add to `dvc.yaml` but not into script) -> `make_checkpoint` (add to Python script) -> `signal_file` (add to script in language-agnostic way) -> `live` (integrate dvclive into Python script). The last two steps are interchangeable since they don't build off each other. The `full_pipeline` branch is a variation of `live` but with a more complex pipeline.

shcheklein commented:
@dberenbaum you have just summarized the conversation I had with @iesahin 🤝 

We agreed that I'll invite him next time to the planning and we'll follow the exact same path you described. 

Question: can we still merge this and redo when we have a single repo? Make it a tutorial under the Experiments use case (a bit more hands on) when we have that Use case ready? WDYT @dberenbaum ?

flippedcoder commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,386 @@
+│ │ ╟ c048966           │ 01:30 PM     │    7 │ 0.082436 │ 0.9747 │ 473987 │ 0.001 │ 0            │
+│ │ ╟ 2910f35           │ 01:30 PM     │    6 │  0.08161 │ 0.9749 │ 473987 │ 0.001 │ 0            │
+│ │ ╟ 255d3aa           │ 01:30 PM     │    5 │ 0.088654 │ 0.9716 │ 473987 │ 0.001 │ 0            │
+```
````
The tables should be consistent throughout the tutorial, but I did switch to the `full_pipeline` branch instead of the `live` branch I was using before this big update.

dberenbaum commented:
> Question: can we still merge this and redo when we have a single repo? Make it a tutorial under the Experiments use case (a bit more hands on) when we have that Use case ready? WDYT @dberenbaum ?

👍 I think we can use the existing branch structure in the right order for now to guide the tutorial until we have a better repo. For now, the collaboration can focus on what belongs in each doc so that neither @flippedcoder nor @iesahin are duplicating each other's efforts.

@shcheklein @dmpetrov Let me know what you think.

flippedcoder commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,386 @@
+Since this is the first run of this project with checkpoints enabled, a new
+checkpoint file will be created. Each time an epoch completes, a new checkpoint
+is added to that file. In the code, you'll find the `dvclive.next_step()` method
+in the training epochs. This line is how checkpoints are created.
```
I'll make sure to update this with the modified repo.

iesahin commented:
I have transferred the new example get started repo: https://github.com/iterative/dvc-get-started-mnist

I'll write a script to generate this code similar to the one in [the current GS project](https://github.com/iterative/example-get-started). It will have tags for incrementally developing the project. 

To summarize:

- Previous one wasn't running as is on Katacoda. It required to reduce the size of `data.xml`. I also tested [checkpoints branches](https://github.com/iterative/dvc-checkpoints-mnist) but [PyTorch is a beast that cannot be caged in Katacoda.](https://github.com/iterative/katacoda-scenarios/issues/15#issuecomment-799449679) 

- This new one runs on Katacoda (with a Docker container) just fine. I didn't test but probably it can run on 1GB RAM . We can say it requires 2GB. The current GS project requires 8GB RAM. (But we say it requires 4GB. 😄)

- We needed some well known persistent dataset across the documentation. MNIST is the most well known I think. 

It looks @flippedcoder did a great job writing a new tutorial for the checkpoints. I'll read the PR in more detail. 

I can go on to integrate checkpoints to this new GS project with similar branches to the Checkpoints repository. Currently it uses _implicit checkpoints_ after `dvc exp run`. I didn't use any Dvclive features or `make_checkpoint` in the code without first testing on Katacoda. We can also have separate repositories, branches on the same repository, projects sharing some common code, etc. I'm open to all these options. 

BTW, would you like to have some docker images for `dvc-checkpoints-mnist` repository? @dberenbaum (iterative/dvc-doc-containers#7)

@dmpetrov @shcheklein @jorgeorpinel

dberenbaum commented:
Thanks, @iesahin! Don't worry about merging the repos or doing much with dvc-checkpoints-mnist just yet. Might get confusing if you are working on it while @flippedcoder is developing this doc.

flippedcoder commented:
@iesahin @dberenbaum I made a super simple repo for this tutorial. https://github.com/iterative/checkpoints-tutorial
It's just the `dvc-checkpoints-mnist` repo on the `full_pipeline` branch with a few minor changes.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,425 @@
+
+If you want to update an existing stage, adding the `-f` option to the
+`dvc stage add --checkpoint` command will force an overwrite of the current
+stage and update it with checkpoints enabled.
```
Do we need to explain `-f` option here? 😄  If the pipeline was defined outside of this it might be better to to jump in the description of `dvc stage add`.

An alternative approach - remove DVC pipeline from the repo and do `dvc init & dvc stage add ...` as a part fo the tutorial. It will reduce the amount of code here and the amount of manual work - copy-pasting a command should be easier than edit dvc.yaml. I like this approach slightly more.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,425 @@
+tracking the metrics along with each checkpoint, so we'll need to add a few
+lines of code.
+
+In the `train.py` file, update the following lines of code in the `main` method
```
`import dvclive` is imported but not used :) It is better to remove it and ask a user to import.

dmpetrov replied:
Also, for some reason train.py does not work from outside of dvc (works just fine from DVC):
```bash
$ python train.py
Traceback (most recent call last):
  File "train.py", line 121, in <module>
    main()
  File "train.py", line 97, in main
    mnist_train = torchvision.datasets.MNIST("data")
  File "/Users/dmitry/mlops-sf/dvc-checkpoints-mnist/.venv/lib/python3.7/site-packages/torchvision/datasets/mnist.py", line 88, in __init__
    raise RuntimeError('Dataset not found.' +
RuntimeError: Dataset not found. You can use download=True to download it
```

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,425 @@
+checkpoints.
+
+Having the number of training epochs defined will prevent your training from
+running indefinitely. You can also hit `Ctrl + C` to stop creating new
```
"running indefinitely" actually is 10 minutes limit in the code. the limit can be removed or the sentence clarified.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,425 @@
+running indefinitely. You can also hit `Ctrl + C` to stop creating new
+checkpoints if the number of epochs are not defined.
+
+### Manually adding checkpoints
```
This section can be removed (moved to another place in the docs) to simplify the tutorial.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,425 @@
+indefinitely until you terminate the code in the terminal. `Ctrl + C` will end
+the process._
+
+### Adding checkpoints conditionally
```
This section can be removed to simplify the tutorial.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,425 @@
+```
+
+Once your epochs are finished, it's time to take a look at the metrics we're
+working with.
````
Terminating is a more common scenario. It might be better to ask users to terminate it after a few epochs (and increase\remove the epochs limit).

dmpetrov commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,425 @@
+```
+
+You can run the following command to promote your experiments to the Git
+workspace.
````
"Git workspace" is not the best term. It should be like "to save your experiments to the Git
history."

dmpetrov reviewed (changes requested):
Great progress, @flippedcoder!
The new repo improved it a lot.

To me it seems like the last polishing iteration is required:
1. I'd suggest removing dvc pipeline (and all dvc related stuff form the repo and code) from the repo (to make it a regular ML project) and run `dvc stage add` instead. 
2. Get rid of "Manually adding checkpoints" & "Condition..." - these two do not bring much value
3. Should we start training with an extremely small `lr` like 0.000001 to show slow progress and more natural metrics dynamics?
4. A could small comments.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+
+Next we'll add a stage to our pipeline to run our download the data we need for
+training our model. In your terminal, run
+`dvc stage add -n download -d download.py -o data/MNIST python download.py`.
```
This command is not needed since `train.py` does the downloading. So, you can significantly simplify this section and combine it with the next one "Enabling DVC pipelines for checkpoint experiments".

Only two simple code blocks are needed to build the entire DVC pipeline:
```bash
$ dvc init
$ dvc stage add -n train -d data/MNIST -d train.py -c model.pt --plots-no-cache predictions.json -p seed,lr,weight_decay --live dvclive python train.py
``` 
and 
```
$ git add .
$ git commit -m 'DVC pipeline'
```

One of the goals of this section should be to show how easy to make a DVC project with checkpoints from a regular ML repository.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+add our checkpoints to give us the points in time with our model that we can
+switch between.
+
+To run this code, run the `dvc exp run` command and it will start training the
```
Also, it might be better to introduce the code changes before running `dvc exp run`. Otherwise, a user won't see the checkpoints.

dmpetrov replied:
The next section is "Running experiments". So, you can just remove `dvc exp run` from the current one. Otherwise, users will try to run it and will be confused by the result.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+## Running experiments
+
+With checkpoints enabled and working in our code, let's run the experiment
+again. You can run an experiment with the `dvc exp run` command. Since we've
```
`dvc exp run` is an action and should be in a new line
```
$ dvc exp run
```

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+## Viewing checkpoints
+
+You can see a table of your experiments and checkpoints in the terminal by
+running `dvc exp show`. The accuracy on our current checkpoints will be high
```
in a new line code block

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+command.
+
+If you want to start a new experiment based on an existing checkpoint, you can
+run `dvc exp apply b3de55f && dvc exp run` where _b3de55f_ is the id of the
```
cmd in a new line

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+checkpoint you want to reference.
+
+You'll be able to see where the experiment starts from the existing checkpoint
+with the `dvc exp show` command. You should seem something similar to this in
```
`dvc exp show` in a new line / code block

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+
+When you've run all the experiments you want to and you are ready to compare
+metrics between checkpoints, you can run the command:
+`dvc metrics diff exp-3c7a8 c23a6e0`. You'll see something similart to this in
```
new line

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,388 @@
+### Looking at plots
+
+You also have the option to generate plots to visualize the metrics about your
+training epochs. Running `dvc plots diff c60fe61 a4d64fc`, where `c60fe61` and
```
new line

dmpetrov reviewed (changes requested):
Good! A few comments inlline. The major ones:
1. I should have been probably noticing this before. It is better to extract in coding blocks (\`\`\` code \`\`\`) all the commands and actions that a user suppose to run. So, it is easy to understand what needs to be copy-pasted and run.
2. The pipeline creation part can be significantly simplified.

dmpetrov commented:
@jorgeorpinel @dberenbaum it looks like the change is close to the final version. I'd appreciate it if you can take a look.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,423 @@
+do that with the following command:
+
+```bash
+dvc stage add -n train -d data/MNIST -d train.py -c model.pt --plots-no-cache predictions.json -p seed,lr,weight_decay --live dvclive python train.py
````
are line separations needed? (I'm asking because I don't see the final result\deployment)

```
dvc stage add -n train -d data/MNIST -d train.py \
            -c model.pt --plots-no-cache predictions.json \
            -p seed,lr,weight_decay --live dvclive python train.py
```

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+## Setting up a DVC pipeline
+
+If you don't have a DVC pipeline setup in your project, adding one only takes a
+few commands. At the root of the project, run:
```
Some motivation for DVC pipeline would be helpful. Like `DVC versions data and models and it also can version the ML model weights file (the checkpoints) during the training. To enable this you will need to run your model training under DVC pipeline. Adding DVC pipline only takes a few commands.`

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+_params.yaml_), some configurations for our plots and showing the training
+metrics.
+
+We'll manually add a slight modification to the way our plots are handled. Add
```
We don't use the confusion matrix in this tutorial, do we? It is better to remove this part and especially yaml modifications (it is not easy to do for users).

Actually the entire section "Enabling DVC pipelines for checkpoint experiments" can be removed. If some explanation is really needed - it can be moved to the previous pipeline creation section.

dmpetrov replied:
I'd suggest removing the confusion matrix plot from the repository and code. It should be a subject of a different tutorial.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+
+Take a look at the _train.py_ file and you'll see how we train a convolutional
+neural network to classify handwritten digits. The main area of this code most
+relevant to checkpoints is when we iterate over the training epochs.
```
A code snippet might be helpful:
```python
        # Iterate over training epochs.
        for i in range(1, EPOCHS+1):
            # Train in batches.
            train_loader = torch.utils.data.DataLoader(
                    dataset=list(zip(x_train, y_train)),
                    batch_size=512,
                    shuffle=True)
            for x_batch, y_batch in train_loader:
                train(model, x_batch, y_batch, params["lr"], params["weight_decay"])
            torch.save(model.state_dict(), "model.pt")
            # Evaluate and checkpoint.
            metrics = evaluate(model, x_test, y_test)
            for k, v in metrics.items():
                print('Epoch %s: %s=%s'%(i, k, v))
```

dmpetrov replied:
Also, you can mention that the line `torch.save(model.state_dict(), "model.pt")` updates the checkpoint file.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+
+By adding the two `dvclive` methods to our training epoch loop, we are enabling
+checkpoints in our code and recording the training metrics at each of those
+checkpoints.
```
A bit more details can be helpful: "To be more specific, `dvclive.log(k, v)` stores metric `k` with a value `v` in plain text files (in `dvclive/` directory by default) and `dvclive.next_step()` signals DVC that it can take a snapshot of the entire workspace and version it with Git. It is important the with this approach only code with meta information is versioned in Git (as an ephemeral commit) while the actual model weight file will be stored in DVC data cache."

dmpetrov commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,423 @@
+Updating lock file 'dvc.lock'
+Checkpoint experiment iteration '77dc46c'.
+Updating lock file 'dvc.lock'
+```
````
I have a slightly different output format. Users can be confused by the difference.
```bash
.....
Epoch 2: loss=1.25374174118042
Epoch 2: acc=0.7738
Updating lock file 'dvc.lock'
Checkpoint experiment iteration '4505b31'.

file:///Users/dmitry/src/tmp/checkpoints-tutorial/dvclive.html
Epoch 3: loss=0.7242147922515869
Epoch 3: acc=0.8284
Updating lock file 'dvc.lock'
Checkpoint experiment iteration '013ea5b'.

file:///Users/dmitry/src/tmp/checkpoints-tutorial/dvclive.html
Epoch 4: loss=0.5083536505699158
Epoch 4: acc=0.8538
Updating lock file 'dvc.lock'
Checkpoint experiment iteration '7a62f2a'.
```

dmpetrov commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+`Ctrl + C`. Now it's time to take a look at the metrics we're working with.
+
+_If you don't have a number of training epochs defined and you don't terminate
+the process, the experiment will run for 10 minutes._
```
Not 10 minutes. 10 epochs instead.

dmpetrov commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,423 @@
+run:
+
+```bash
+dvc exp apply b3de55f && dvc exp run
````
`dvc exp apply b3de55f && dvc exp run` - this action is not very useful. Users will get the same result.
We should do some code modification (like increase learning rate) before `dvc exp run`.

dmpetrov reviewed (changes requested):
@flippedcoder good change!
I added a few more comments. Mostly minor.

dberenbaum reviewed (commented):
Looking good! I think the overall workflow makes sense.

I have a couple general thoughts:

* We might need some branch, tag, or other version of the repo that show the final results, or at least the `dvc.yaml` and the updates to `train.py` in case people get stuck trying to make those changes manually.
* This doesn't cover the other scenarios from the branches in the original checkpoints repo. I think excluding those is a good idea, but we probably need another ticket to put all of those scenarios somewhere in the user guide. cc @jorgeorpinel

dberenbaum commented on content/docs/sidebar.json:
```diff
@@ -135,6 +135,12 @@
         ]
       },
       "experiment-management",
+      {
+        "slug": "checkpoints",
+        "tutorials": {
+          "katacoda": "https://katacoda.com/dvc/courses/get-started/checkpoints"
+        }
+      },
```
You can preview the deployment now: https://dvc-org-mm-checkpoints--lv46zp.herokuapp.com/doc/user-guide/checkpoints! It looks like the navigation here needs to be modified since it's not showing up as a subsection of experiment management now.

flippedcoder replied:
Just finished that round of updates! Let me know what y'all think! @dberenbaum @dmpetrov @jorgeorpinel

flippedcoder commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+- Track all code and data changes corresponded to the checkpoints.
+- Return a report when metrics start diverging.
+
+You can also automate the process of tracking every training epoch, return to
+any previous checkpoint, see when metrics converge or when the learning rate
+needs to be adjusted, and even recover model weights, parameters, and code. When
+you adjust parameters and code, DVC tracks those changes for you without adding
+a lot of commits to your workspace.
```
Yeah, it's more of an expansion on those details. I can merge this with the bullets if that makes more sense.

flippedcoder commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+_params.yaml_), some configurations for our plots and showing the training
+metrics.
+
+We'll manually add a slight modification to the way our plots are handled. Add
```
Honestly, I just didn't know how to include that in the `dvc stage add` command. 😅 If that's possible, I'll add it to that part. Otherwise I'll use `dvc plots modify`.

dberenbaum replied:
Yes, unfortunately it cannot be added in `dvc stage add`!

flippedcoder commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+
+## Running experiments
+
+With checkpoints enabled and working in our code, let's run the experiment
+again. You can run an experiment with the following command:
```
Good catch! Just some leftover text from earlier versions.

flippedcoder commented on content/docs/user-guide/checkpoints.md:
````diff
@@ -0,0 +1,423 @@
+dvc exp show
+```
+
+The accuracy on our current checkpoints will be high because we've already run
+them twice and training always picks up from the last checkpoint.
````
Yep, just some text from an earlier version. :)

jorgeorpinel commented:
> we probably need another ticket to put all of those scenarios somewhere in the user guide

Makes sense @dberenbaum want to create it? I'm shifting gears a little from 2.0-related things into other docs directions currently, but having that issue in for now should help us get to it at some point in the future. Thanks

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,409 @@
+
+## Setting up a DVC pipeline
+
+If you don't have a DVC pipeline setup in your project, adding one only takes a
```
It is still not clear why user needs a DVC pipeline. You can find it here: https://github.com/iterative/dvc.org/pull/2373#discussion_r622459438

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,409 @@
+following command:
+
+```bash
+dvc stage add -n train -d data/MNIST -d train.py -c model.pt --plots-no-cache predictions.json -p seed,lr,weight_decay --live dvclive python train.py
````
New lines are needed like:
```bash
dvc stage add -n train -d data/MNIST -d train.py \
                             -c model.pt --plots-no-cache predictions.json
                             -p seed,lr,weight_decay --live dvclive python train.py
```

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,409 @@
+git commit -m "created DVC pipeline"
+```
+
+## Enabling DVC pipelines for checkpoint experiments
````
I'd suggest moving it right after the command run `dvc stages add` and before Git commit. Otherwise, it looks for a user as a separate action.

The explanation can be reduced a bit.

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,409 @@
+
+By adding the two `dvclive` methods to our training epoch loop, we are enabling
+checkpoints in our code and recording the training metrics at each of those
+checkpoints.
```
A bit more details might be helpful: https://github.com/iterative/dvc.org/pull/2373#discussion_r622482174

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,409 @@
+
+where _b3de55f_ is the id of the checkpoint you want to reference.
+
+Next, we'll change the learning rate in the _params.yaml_ to `0.001`. Then we'll
```
just an idea:
```
dvc exp run --set-param lr=0.001
``` 
might be an easier way (for user) of doing it.

flippedcoder replied:
@dmpetrov @dberenbaum I tried running this command in few different ways and I get this error each time:

```
Traceback (most recent call last):
  File "/Users/milecia/Repos/checkpoints-tutorial/train.py", line 123, in <module>
    main()
  File "/Users/milecia/Repos/checkpoints-tutorial/train.py", line 110, in main
    train(model, x_batch, y_batch, params["lr"], params["weight_decay"])
  File "/Users/milecia/Repos/checkpoints-tutorial/train.py", line 45, in train
    optimizer = torch.optim.Adam(model.parameters(), lr=lr,
  File "/Users/milecia/Repos/checkpoints-tutorial/.venv/lib/python3.9/site-packages/torch/optim/adam.py", line 36, in __init__
    if not 0.0 <= lr:
TypeError: '<=' not supported between instances of 'float' and 'str'
```

Any thoughts or suggestions?

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,409 @@
+│ │ ╟ b3de55f           │ 05:02 PM     │    4 │ 0.41666  │ 0.8777 │ 473987 │ 0.0001 │ 0            │
+```
+
+The existing checkpoint is referenced at the beginning of the new experiment.
````
The sha-s would be super helpful here. It might be not clear for the user what is existing and a new mean.

flippedcoder replied:
What do you mean by the 'sha-s'?

dberenbaum replied:
@flippedcoder This means not using the `exp-4de2a` names but the Git commit hashes instead. I think if you upgrade to the latest dvc version, you will be able to see both and it will fix this issue.

flippedcoder replied:
Gotcha. I think I have the latest DVC version. Is it `2.0.18`? I still don't see the Git hashes though. Just the `exp-f923ji` syntax.

dberenbaum replied:
It's 2.1.0 now. The table format hopefully looks something like:
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┓
┃ Experiment              ┃ Created      ┃ step ┃ acc    ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━┩
│ workspace               │ -            │ 1    │ 0.9133 │
│ test                    │ 12:52 PM     │ 1    │ 0.9133 │
│ live                    │ Apr 19, 2021 │ -    │ -      │
│ │ ╓ 7243788 [exp-bffd1] │ 12:46 PM     │ 1    │ 0.9133 │
│ ├─╨ c039cb0             │ 12:46 PM     │ 0    │ 0.8523 │
│ │ ╓ 9121bba [exp-dbb5a] │ 12:46 PM     │ 6    │ 0.9791 │
│ │ ╟ 9790bd2             │ 12:45 PM     │ 5    │ 0.9807 │
│ │ ╟ 5d629a2             │ 12:45 PM     │ 4    │ 0.976  │
│ │ ╟ b94ecb0             │ 12:45 PM     │ 3    │ 0.9714 │
│ │ ╟ 5a7ba19             │ 12:45 PM     │ 2    │ 0.9569 │
│ │ ╟ 0849314             │ 12:45 PM     │ 1    │ 0.9659 │
│ ├─╨ 98c13ed             │ 12:44 PM     │ 0    │ 0.95   │
└─────────────────────────┴──────────────┴──────┴────────┘
```

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,409 @@
+dvclive.json  acc       0.9854   0.9831   -0.0023
+dvclive.json  loss      0.04828  0.05095  0.00267
+dvclive.json  step      19       15       -4
+```
````
You can reference that these are the same numbers as you see in the table but in a different format.

dmpetrov commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,409 @@
+        model.pt
+        plots.html
+        predictions.json
+```
````
It looks like a commit is missing: 
```
git commit -m 'my experiment'
```

and, probably a last "motivational" sentence about checkpoint 😄

dmpetrov reviewed (approved):
Great doc @flippedcoder!

A few minor comments are inline. I hope @jorgeorpinel & @dberenbaum will have a chance to review and approve it as well.

dberenbaum commented:
@jorgeorpinel Do you think https://dvc.org/doc/user-guide/experiment-management#checkpoints-in-source-code needs to be updated to reflect this doc?

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,454 @@
+              -p seed,lr,weight_decay --live dvclive python train.py
+```
+
+The checkpoints need to be enabled in DVC at the pipeline level. The
+`-c / --checkpoint` option of the `dvc stage add` command defines the checkpoint
+file or directory.
````
I'm not sure it's clear why `model.pt` is the checkpoint file here and not `predictions.json` or some other output. It might be helpful to clarify that the checkpoint file is an output from one checkpoint that becomes a dependency for the next checkpoint, such as the model weights file. cc @jorgeorpinel @pmrowla

jorgeorpinel replied:
Agree, introduce "circular dependency" concept (not sure about the terminology though)

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,454 @@
+do that with `dvc stage add`, which we'll explain more later. For now, run the
+following command:
+
+```bash
+dvc stage add -n train -d data/MNIST -d train.py \
+              -c model.pt --plots-no-cache predictions.json \
+              -p seed,lr,weight_decay --live dvclive python train.py
+```
````
Thoughts on using long flag names (`--name`) and using one flag per line to make this more readable? cc @dmpetrov @jorgeorpinel

jorgeorpinel replied:
Agree to use full option names and one arg per line at least the first time they're shown, but only the ones related to the topic, I think. I think it's OK to keep `-n train` along with the `-d ...` in the first line though.

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,454 @@
+
+```yaml
+stages:
+  download:
+    cmd: python download.py
+    deps:
+      - download.py
+    outs:
+      - data/MNIST
````
Where did this stage come from? I don't see it in the repo or the above instructions. Is it from an old version?

flippedcoder replied:
Oh yeah. That was from an old version.

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,454 @@
+        print('Epoch %s: %s=%s'%(i, k, v))
+```
+
+The line `torch.save(model.state_dict(), "model.pt")` updates the checkpoint
+file.
````
This isn't an update to the existing code, though, right? I'm not sure this line is needed, or maybe it belongs somewhere else.

flippedcoder replied:
Right, it's not an update to the code. It's just an explanation of where checkpoints are created in the code. I added it per this comment https://github.com/iterative/dvc.org/pull/2373#discussion_r622475955

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,454 @@
+    # Evaluate and checkpoint.
+    metrics = evaluate(model, x_test, y_test)
+    for k, v in metrics.items():
+        print('Epoch %s: %s=%s'%(i, k, v))
```
It seems like this code block is missing the `dvclive` additions?

dberenbaum replied:
You could change the code format from `python` to `diff` and do:

```suggestion
-        print('Epoch %s: %s=%s'%(i, k, v))
+        dvclive.log(k, v)
+    dvclive.next_step()
```

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,454 @@
+Checkpoint experiment iteration '69beaaa'.
+```
+
+After a few epochs have completed, stop terminate the training process with
````
```suggestion
After a few epochs have completed, stop the training process with
```

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,454 @@
+│ │ ╟ 868c144           │ 04:37 PM     │   17 │ 0.12042  │ 0.9616 │ 473987 │ 0.001  │ 0            │
+│ │ ╟ 4de4e69           │ 04:37 PM     │   16 │ 0.13808  │ 0.9585 │ 473987 │ 0.001  │ 0            │
+│ │ ╟ d8d633c           │ 04:37 PM     │   15 │ 0.20415  │ 0.9397 │ 473987 │ 0.001  │ 0            │
+│ │ ╟ 8fbab93 (b3de55f) │ 04:36 PM     │   14 │ 0.24509  │ 0.9309 │ 473987 │ 0.001  │ 0            │
```
The `step` value here should be 5 based on `b3de55f`. Can you try with the most recent version of dvc and dvclive? If you still have this issue, it might be a bug in dvc or dvclive. cc @pmrowla @pared

dberenbaum replied:
It also doesn't seem like the model weights were applied correctly here since it's unlikely they jumped from 0.8777 in b3de55f to 0.9309 in 8fbab93.

pared replied:
@flippedcoder Did you try with the latest version? Is the problem still occuring?

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,454 @@
+generate a `plots.html` file that you can open in a browser and it will display
+plots for you similar to the ones below.
+
+![Plots generated from running experiments on MNIST dataset using DVC](/img/checkpoints_plots.png)
+_The results here won't show anything interesting because the model increases in
+accuracy so fast._
```
It seems like this plot is not only uninteresting, but also kind of confusing since there's only one line visible on it despite comparing two checkpoints. If you compare the last checkpoints in both the original and the modified experiment, you should end up with something more interesting since the different learning rates should cause them to diverge.

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,454 @@
+new set of checkpoints under a new experiment branch.
+
+```dvc
+┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━┳━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━┓
+┃ Experiment            ┃ Created  ┃ step ┃    loss ┃    acc ┃ seed   ┃ lr     ┃ weight_decay ┃
+┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━╇━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━┩
+│ workspace             │ -        │    4 │ 0.41666 │ 0.8777 │ 473987 │ 0.0001 │ 0            │
+│ main                  │ 04:49 PM │    - │       - │      - │ 473987 │ 0.001  │ 0            │
+│ │ ╓ exp-a75bb         │ 05:18 PM │    4 │ 0.41666 │ 0.8777 │ 473987 │ 0.0001 │ 0            │
+│ │ ╟ b8a4ceb           │ 05:18 PM │    3 │ 0.50835 │ 0.8538 │ 473987 │ 0.0001 │ 0            │
+│ │ ╟ 7b652b1           │ 05:17 PM │    2 │ 0.72421 │ 0.8284 │ 473987 │ 0.0001 │ 0            │
+│ │ ╟ cf0168f           │ 05:17 PM │    1 │  1.2537 │ 0.7738 │ 473987 │ 0.0001 │ 0            │
+│ ├─╨ 8c7f40f           │ 05:17 PM │    0 │  1.9428 │ 0.5715 │ 473987 │ 0.0001 │ 0            │
+│ │ ╓ exp-7fafd         │ 05:10 PM │    5 │ 0.36601 │ 0.8943 │ 473987 │ 0.0001 │ 0            │
+│ │ ╟ 9ac9e6a           │ 05:09 PM │    4 │ 0.41666 │ 0.8777 │ 473987 │ 0.0001 │ 0            │
+│ │ ╟ b869efe           │ 05:09 PM │    3 │ 0.50835 │ 0.8538 │ 473987 │ 0.0001 │ 0            │
+│ │ ╟ 422395e           │ 05:09 PM │    2 │ 0.72421 │ 0.8284 │ 473987 │ 0.0001 │ 0            │
+│ │ ╟ 44e44a9           │ 05:09 PM │    1 │  1.2537 │ 0.7738 │ 473987 │ 0.0001 │ 0            │
+│ ├─╨ c6e11b4           │ 05:09 PM │    0 │  1.9428 │ 0.5715 │ 473987 │ 0.0001 │ 0            │
````
Seems like this example is from a different set of experiments. It probably makes sense to clean these up and reproduce all the experiment results once everything else is pretty much resolved.

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,454 @@
+│ ├─╨ c6e11b4           │ 05:09 PM │    0 │  1.9428 │ 0.5715 │ 473987 │ 0.0001 │ 0            │
+```
+
+We can also remove all of the experiments we don't promote to our Git workspace
+with the following command:
+
+```bash
+dvc exp gc --workspace --force
+```
````
I'd suggest moving this to the bottom of the next section since it won't do anything if the users run it at this point before committing any experiment results to Git.

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,454 @@
+
+To promote an experiment to a Git branch run:
+
+        dvc exp branch <exp>
```
If you want to keep all of the checkpoints as commits in your Git history, you can use `dvc exp branch` to do that. I'm torn on whether it's worthwhile to explain that here. cc @dmpetrov @jorgeorpinel

jorgeorpinel replied:
I wouldn't mention `exp branch` out of context here. Actually I'd skip the whole code block above, as the `git` command is repeated below.

dberenbaum reviewed (commented):
Great stuff @flippedcoder! The main changes I see at this point are making sure everything works as expected with your code examples so that you have clean, consistent output throughout the tutorial.

jorgeorpinel commented on content/docs/user-guide/checkpoints.md:
```diff
@@ -0,0 +1,423 @@
+[The way checkpoints are implemented by DVC](/blog/experiment-refs) utilizes
+_ephemeral_ experiment commits and experiment branches within DVC. They are
+created using the metadata from experiments and are tracked with the `exps`
+custom Git reference.
```
Sorry guys this review is really long and I didn't want to further complicate the process. I think when you approve we should merge and then I can make a follow up with copy edits for consistency with existing terminology etc.

For example: we didn't end up using the term "ephemeral experiment". Instead we talk about making experiments persistent (by applying them + committing them to Git).

jorgeorpinel commented:
> Do you think https://dvc.org/doc/user-guide/experiment-management#checkpoints-in-source-code needs to be updated to reflect this doc?

It should probably link to this doc, yep! It can be part of a (possible) follow up copy edit PR if needed, or feel free to include it, @flippedcoder .

dberenbaum commented:
> > Do you think https://dvc.org/doc/user-guide/experiment-management#checkpoints-in-source-code needs to be updated to reflect this doc?
> 
> It should probably link to this doc, yep! It can be part of a (possible) follow up copy edit PR if needed, or feel free to include it, @flippedcoder .

Sounds good! Let's leave it for the copy edit PR then. There's enough going on in this one already.

jorgeorpinel commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,456 @@
+With checkpoints enabled and working in our code, let's run the experiment. You
+can run an experiment with the following command:
+
+```bash
+dvc exp run
````
Please use the `dvc` highlighter (instead of `bash`) and `$` prompts throughout. Here's one:
````suggestion
```dvc
$ dvc exp run
````

jorgeorpinel commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,456 @@
+You'll see output similar to this in your terminal while the training process is
+going on.
+
+```dvc
````
In its current form (no commands) this one doesn't need highlighting:
````suggestion
```
````

jorgeorpinel commented on content/docs/user-guide/experiment-management/checkpoints.md:
````diff
@@ -0,0 +1,456 @@
+
+You should see something similar to this in your terminal.
+
+```git
````
Same for this one and any other that's just terminal output:
````suggestion
```
````

jorgeorpinel reviewed (approved):
Please ping me if I don't notice when this gets merged and I'll do a copy edit follow up.

flippedcoder commented:
Alright! I think I've addressed all of the comments. I tried my best not to miss anything. 😅 It should be ready to merge, but let me know if y'all have any other feedback! @dmpetrov @dberenbaum @jorgeorpinel

dberenbaum commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -0,0 +1,478 @@
+custom Git reference.
+
+You can add experiments to your Git history by committing the experiment you
+want to track, which you'll see later in this t$ utorial.
```
```suggestion
want to track, which you'll see later in this tutorial.
```

dberenbaum reviewed (approved):
LGTM! 🎉
